ykrt/src/trace/swt/mod.rs

```diff
@@ -73,77 +100,47 @@ struct SWTTraceRecorder {}
-                        len: u64::try_from(bc_section.len()).unwrap(),
-                    };
-                    unsafe { get_function_names(bs, &mut functions, &mut functions_len) };
-                    for entry in unsafe { std::slice::from_raw_parts(functions, functions_len) } {
-                        fnames.borrow_mut().insert(
-                            entry.index,
-                            unsafe { std::ffi::CStr::from_ptr(entry.name) }.to_owned(),
-                        );
-                    }
-                });
-
-                aot_blocks.extend(tb.borrow().iter().map(|tb| {
-                    match fnames.borrow_mut().get(&tb.function_index) {
-                        Some(name) => TraceAction::MappedAOTBlock {
-                            func_name: name.to_owned(),
-                            bb: tb.block_index,
-                        },
-                        _ => panic!(
-                            "Failed to get function name by index {:?}",
-                            tb.function_index
-                        ),
-                    }
-                }));
-            })
-        });
-        if aot_blocks.len() > DEFAULT_TRACE_TOO_LONG {
+        let bbs = BASIC_BLOCKS.with(|tb| tb.replace(Vec::new()));
+        if bbs.len() > DEFAULT_TRACE_TOO_LONG {
             Err(TraceRecorderError::TraceTooLong)
-        } else if aot_blocks.is_empty() {
+        } else if bbs.is_empty() {
             // FIXME: who should handle an empty trace?
             panic!();
         } else {
-            Ok(Box::new(SWTraceIterator::new(aot_blocks)))
+            Ok(Box::new(SWTraceIterator::new(bbs)))
         }
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
-
-    #[test]
-    fn test_basic_blocks_content() {
-        assert_eq!(BASIC_BLOCKS.with(|blocks| blocks.borrow().clone()), []);
-        trace_basicblock(0, 0);
-        trace_basicblock(0, 1);
-        trace_basicblock(1, 0);
-        assert_eq!(
-            BASIC_BLOCKS.with(|blocks| blocks.borrow().clone()),
-            vec![
-                TracingBlock {
-                    function_index: 0,
-                    block_index: 0,
-                },
-                TracingBlock {
-                    function_index: 0,
-                    block_index: 1,
-                },
-                TracingBlock {
-                    function_index: 1,
-                    block_index: 0,
-                },
-            ]
-        );
+struct SWTraceIterator {
+    bbs: std::vec::IntoIter<TracingBlock>,
+}
+
+impl SWTraceIterator {
+    fn new(bbs: Vec<TracingBlock>) -> SWTraceIterator {
+        return SWTraceIterator {
+            bbs: bbs.into_iter(),
+        };
+    }
+}
+
+impl Iterator for SWTraceIterator {
+    type Item = Result<TraceAction, AOTTraceIteratorError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.bbs
+            .next()
+            .map(|tb| match FUNC_NAMES.get(&tb.function_index) {
```

Comment:
Just to make sure I understand how LazyLock works - is the first lazy initialisation of FUNC_NAMES will be on this line?

Reply:
Yes (meaning: it happens in the compiler thread).

Reply:
Will there be any benefit in moving this initialisation before tracing starts?

Reply:
>  Will there be any benefit in moving this initialisation before tracing starts?

Yes, possibly! But I haven't benchmarked it (yet) so decided not to. But conceptually we would do `thread::spawn(|| FUNC_NAMES.get())` in `SWTracer::new`. I would only do that if it shows up as a noticable cost, though, as creating threads isn't cheap either.